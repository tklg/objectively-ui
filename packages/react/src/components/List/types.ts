import { ForwardRefExoticComponent } from 'react'
import { PropsWithChildrenAndClassName } from 'src/types/PropsWithChildrenAndClassName'

export interface ListProps extends PropsWithChildrenAndClassName {
  ref?: React.RefObject<HTMLUListElement>;
  compact?: boolean;
  divided?: boolean;
  noPadding?: boolean;
}

export interface ListContextProps {
  compact: boolean;
  divided: boolean;
}

export type ListComponent<P = ListProps> = ForwardRefExoticComponent<P>
