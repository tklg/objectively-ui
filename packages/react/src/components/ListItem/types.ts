import { NamedExoticComponent } from 'react'
import { PropsWithChildrenAndClassName } from 'src/types/PropsWithChildrenAndClassName'

export interface ListItemProps extends PropsWithChildrenAndClassName {
  description?: string;
  selected?: boolean;
  button?: boolean;
}

export type ListItemComponent<P = object> = NamedExoticComponent<P>
