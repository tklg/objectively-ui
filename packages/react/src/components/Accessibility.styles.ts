import { ColorTheme } from 'src/types/ColorTheme'

export const focusOutlineStylesWithoutElement = (theme: ColorTheme, offset?: number) => ({
  outline: `2px solid ${theme.colors.focusOutline}`,
  outlineOffset: offset ?? 2,
  borderRadius: 3,
})

export const focusOutlineStyles = (theme: ColorTheme, offset?: number) => ({
  '&:focus-visible': focusOutlineStylesWithoutElement(theme, offset),
})

export const focusOutlineBeforeStyles = (theme: ColorTheme, offset?: number) => ({
  '&:focus-visible:before': {
    content: '\'\'',
    position: 'absolute',
    height: '100%',
    width: '100%',
    top: 0,
    left: 0,
    ...focusOutlineStylesWithoutElement(theme, offset),
  },
})
