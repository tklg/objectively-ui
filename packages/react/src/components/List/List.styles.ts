import { css } from '@emotion/react'
import { ColorTheme } from 'src/types/ColorTheme'
import { PROJECT_SHORTNAME } from 'src/utils/constants'

export const listStyles = (theme: ColorTheme) => css({
  listStyle: 'none',
  margin: 0,
  padding: `${theme.spacing.sm} 0`,

  [`&.${PROJECT_SHORTNAME}-List-noPadding`]: {
    padding: 0,
  },
})
