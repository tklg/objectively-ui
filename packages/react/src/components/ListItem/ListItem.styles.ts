import { css } from '@emotion/react'
import { focusOutlineStylesWithoutElement } from 'src/components/Accessibility.styles'
import { ColorTheme } from 'src/types/ColorTheme'
import { PROJECT_SHORTNAME } from 'src/utils/constants'

export const listItemStyles = (theme: ColorTheme) => css({
  padding: 0,
  height: theme.size.md,
  lineHeight: theme.size.md,
  color: theme.colors.textPrimary,

  [`&.${PROJECT_SHORTNAME}-ListItem-divided`]: {
    '&:not(:first-of-type)': {
      borderTop: `${theme.lines.xs} solid ${theme.colors.divider}`,
    },
  },

  [`&.${PROJECT_SHORTNAME}-ListItem-compact`]: {
    height: theme.size.sm,
    lineHeight: theme.size.sm,
  },
})

export const listItemButtonStyles = (theme: ColorTheme) => css({
  display: 'block',
  border: 'none',
  height: '100%',
  width: '100%',
  textAlign: 'left',
  background: 'transparent',
  padding: 0,
  margin: 0,
  color: theme.colors.textPrimary,
  cursor: 'pointer',

  '&:focus-visible': {
    outline: 'none',
  },

  '&:disabled': {
    color: theme.colors.textDisabled,
    cursor: 'default',
  },
})

export const listItemContentStyles = (theme: ColorTheme) => css({
  position: 'relative',
  padding: `0 ${theme.spacing.md}`,
  height: '100%',
  width: '100%',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  transition: [
    `background ${theme.transitions.duration.fast} ${theme.transitions.function.default}`,
  ].join(','),

  'button:not(:disabled) &:hover': {
    background: theme.colors.hover,
  },
  'button:not(:disabled) &:active': {
    background: theme.colors.accentPrimary.shadow,
  },
  'button &:before': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    height: '100%',
    width: '100%',
  },
  'button:focus-visible &:before': focusOutlineStylesWithoutElement(theme),

  '&[data-selected=true]': {
    background: theme.colors.accentPrimary.value,
    color: theme.colors.accentPrimary.contrastText,

    '&:hover': {
      background: theme.colors.accentPrimary.hover,
    },
    '&:active': {
      background: theme.colors.accentPrimary.active,
    },
  },

  [`.${PROJECT_SHORTNAME}-ListItem-joined &`]: {
    margin: `2px ${theme.spacing.xxs}`,
    borderRadius: theme.radii.sm,
    width: `calc(100% - (${theme.spacing.xxs} * 2))`,
    height: 'calc(100% - 4px)',
    padding: `0 calc(${theme.spacing.md} - ${theme.spacing.xxs})`,
  },
})

export const listItemTextStyles = (theme: ColorTheme) => css({
  display: 'block',
  overflowX: 'hidden',
  whiteSpace: 'nowrap',
  textOverflow: 'ellipsis',
})

export const listItemSubtextStyles = (theme: ColorTheme) => css({
  display: 'block',
  overflowX: 'hidden',
  whiteSpace: 'nowrap',
  textOverflow: 'ellipsis',
  fontSize: theme.typography.size.xs,
  marginTop: -2,
  color: theme.colors.textSecondary,

  '[data-selected=true] &': {
    color: theme.colors.accentPrimary.contrastTextSecondary,
  },
  [`.${PROJECT_SHORTNAME}-ListItem-compact &`]: {
    marginTop: -4,
  },
})
