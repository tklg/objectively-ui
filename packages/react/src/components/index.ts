export * from './AppLayout'
export * from './Button'
export * from './CssBaseline'
export * from './Heading'
export * from './Input'
export * from './SideNavigation'
export * from './TopNavigation'
