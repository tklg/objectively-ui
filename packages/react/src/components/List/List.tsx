import { createContext, forwardRef, useMemo } from 'react'
import { listStyles } from 'src/components/List/List.styles'
import { ListComponent, ListContextProps, ListProps } from 'src/components/List/types'
import { useTheme } from 'src/hooks'
import { buildClassName } from 'src/utils/buildClassName'

const ELEMENT_NAME = 'List'

export const ListContext = createContext<ListContextProps>({
  compact: false,
  divided: false,
})

export const List = forwardRef<HTMLUListElement, ListProps>(({
  children,
  className: _className,
  compact,
  divided,
  noPadding,
}, ref) => {
  const theme = useTheme()
  const className = buildClassName(ELEMENT_NAME, {
    compact,
    divided,
    noPadding,
  }, _className)

  const contextValue = useMemo(() => ({
    compact: compact ?? false,
    divided: divided ?? false,
  }), [compact])

  return (
    <ListContext.Provider value={contextValue}>
      <ul
        ref={ref}
        className={className}
        css={listStyles(theme)}
      >
        {children}
      </ul>
    </ListContext.Provider>
  )
}) as ListComponent

if (process.env.NODE_ENV !== 'production') {
  List.displayName = ELEMENT_NAME
}
