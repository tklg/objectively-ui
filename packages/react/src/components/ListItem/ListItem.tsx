import { css } from '@emotion/react'
import { forwardRef, useMemo } from 'react'
import { useListContext } from 'src/components/List/useListContext'
import { listItemButtonStyles, listItemStyles, listItemContentStyles, listItemTextStyles, listItemSubtextStyles } from 'src/components/ListItem/ListItem.styles'
import { ListItemComponent, ListItemProps } from 'src/components/ListItem/types'
import { useTheme } from 'src/hooks'
import { buildClassName } from 'src/utils/buildClassName'

const ELEMENT_NAME = 'ListItem'

export const ListItem = forwardRef<HTMLLIElement, ListItemProps>(({
  children,
  description,
  selected,
  button,
  className: _className,
  onClick,
}, ref) => {
  const { divided, compact } = useListContext()
  const theme = useTheme()
  const className = buildClassName(ELEMENT_NAME, {
    selected,
    button,
    divided,
    joined: !divided,
    compact,
    description: Boolean(description),
  }, _className)
  const listItemTextClassName = buildClassName(`${ELEMENT_NAME}Text`)
  const listItemTextDesciptionClassName = buildClassName(`${ELEMENT_NAME}Subtext`)

  const content = useMemo(() => (
    <div
      css={listItemContentStyles(theme)}
      data-selected={selected}
    >
      <div css={css({ width: '100%' })}>
        <span className={listItemTextClassName} css={listItemTextStyles(theme)}>{children}</span>
        {description && <span className={listItemTextDesciptionClassName} css={listItemSubtextStyles(theme)}>{description}</span>}
      </div>
    </div>
  ), [description, children, listItemTextClassName, listItemTextDesciptionClassName, theme])

  return (
    <li
      ref={ref}
      className={className}
      css={listItemStyles(theme)}
    >
      {button ? (
        <button
          css={listItemButtonStyles(theme)}
          onClick={onClick}
        >
          {content}
        </button>
      ) : (
        content
      )}
    </li>
  )
}) as ListItemComponent

if (process.env.NODE_ENV !== 'production') {
  ListItem.displayName = ELEMENT_NAME
}
