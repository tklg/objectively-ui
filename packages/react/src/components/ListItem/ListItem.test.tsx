import { fireEvent, render, screen } from 'src/__test__/render'
import { ListItem } from 'src/components/ListItem'
import { List } from 'src/components/List/List'

describe('ListItem', () => {
  it('renders a list item', async () => {
    render(
      <List>
        <ListItem>List item 1</ListItem>
      </List>,
    )

    expect(screen.getByRole('listitem')).toHaveTextContent('List item 1')
    expect(screen.getByRole('list')).toMatchSnapshot()
  })

  it('renders a list item with a description', async () => {
    render(
      <List>
        <ListItem description='Description'>List item 1</ListItem>
      </List>,
    )

    expect(screen.getByRole('listitem')).toHaveTextContent('List item 1')
    expect(screen.getByRole('listitem')).toHaveTextContent('Description')
    expect(screen.getByRole('list')).toMatchSnapshot()
  })

  it('renders a button list item', async () => {
    const handleClick = jest.fn()

    render(
      <List>
        <ListItem button onClick={handleClick}>List item 1</ListItem>
      </List>,
    )

    fireEvent.click(screen.getByRole('button'))

    expect(handleClick).toHaveBeenCalledTimes(1)
    expect(screen.getByRole('button')).toHaveTextContent('List item 1')
    expect(screen.getByRole('list')).toMatchSnapshot()
  })

  it('renders a button list item with a description', async () => {

    render(
      <List>
        <ListItem description='Description' button onClick={jest.fn()}>List item 1</ListItem>
      </List>,
    )

    expect(screen.getByRole('button')).toHaveTextContent('List item 1')
    expect(screen.getByRole('button')).toHaveTextContent('Description')
    expect(screen.getByRole('list')).toMatchSnapshot()
  })
})
