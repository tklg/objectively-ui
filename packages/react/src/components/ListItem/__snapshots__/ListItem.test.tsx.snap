// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders a button list item 1`] = `
<ul
  class="ObjUI-List emotion-0"
>
  <li
    class="ObjUI-ListItem ObjUI-ListItem-button ObjUI-ListItem-joined emotion-1"
  >
    <button
      class="emotion-2"
    >
      <div
        class="emotion-3"
      >
        <div
          class="emotion-4"
        >
          <span
            class="ObjUI-ListItemText emotion-5"
          >
            List item 1
          </span>
        </div>
      </div>
    </button>
  </li>
</ul>
`;

exports[`ListItem renders a button list item with a description 1`] = `
<ul
  class="ObjUI-List emotion-0"
>
  <li
    class="ObjUI-ListItem ObjUI-ListItem-button ObjUI-ListItem-joined ObjUI-ListItem-description emotion-1"
  >
    <button
      class="emotion-2"
    >
      <div
        class="emotion-3"
      >
        <div
          class="emotion-4"
        >
          <span
            class="ObjUI-ListItemText emotion-5"
          >
            List item 1
          </span>
          <span
            class="ObjUI-ListItemSubtext emotion-6"
          >
            Description
          </span>
        </div>
      </div>
    </button>
  </li>
</ul>
`;

exports[`ListItem renders a list item 1`] = `
<ul
  class="ObjUI-List emotion-0"
>
  <li
    class="ObjUI-ListItem ObjUI-ListItem-joined emotion-1"
  >
    <div
      class="emotion-2"
    >
      <div
        class="emotion-3"
      >
        <span
          class="ObjUI-ListItemText emotion-4"
        >
          List item 1
        </span>
      </div>
    </div>
  </li>
</ul>
`;

exports[`ListItem renders a list item with a description 1`] = `
<ul
  class="ObjUI-List emotion-0"
>
  <li
    class="ObjUI-ListItem ObjUI-ListItem-joined ObjUI-ListItem-description emotion-1"
  >
    <div
      class="emotion-2"
    >
      <div
        class="emotion-3"
      >
        <span
          class="ObjUI-ListItemText emotion-4"
        >
          List item 1
        </span>
        <span
          class="ObjUI-ListItemSubtext emotion-5"
        >
          Description
        </span>
      </div>
    </div>
  </li>
</ul>
`;
